poll/poll.py

```diff
@@ -1236,3 +1265,38 @@ def prepare_data(self):
                         row.append(answers_dict[choice])
                 data[sm.student.id] = row
         return [header_row + questions] + data.values()
+
+    def generate_report_data(self, user_state_iterator, limit_responses=None):
+        """
+        Return a list of student responses to this block in a readable way.
+
+        Arguments:
+            user_state_iterator: iterator over UserStateClient objects.
+                E.g. the result of user_state_client.iter_all_for_block(block_key)
+
+            limit_responses (int|None): maximum number of responses to include.
+                Set to None (default) to include all.
+
+        Returns:
+            each call returns a tuple like:
+            ("username", {
+                           "Question ID": "fun"
+                           "Question": "Are you having fun?"
+                           "Answer": "Yes",
+                           "Answer ID": "1545204793464"
+            })
+        """
+        answers_dict = dict(self.answers)
+        questions_dict = dict(self.questions)
+        for user_state in user_state_iterator:
+            choices = user_state.state['choices']  # {u'enjoy': u'Y', u'recommend': u'N', u'learn': u'M'}
+            for q, a in choices.items():
+                question = questions_dict[q]['label']
+                answer = answers_dict[a]
+                report = {
+                    self.ugettext("Question ID"): q,
+                    self.ugettext("Question"): question,
+                    self.ugettext("Answer ID"): a,
+                    self.ugettext("Answer"): answer
+                }
+                yield (user_state.username, report)
```

Comment:
@xitij2000 Do you mean things like Capa problems that can have multiple questions in a single block? I thought we're handling that already. The capa `generate_report_data ` code definitely [looks like it yields multiple rows per block per student](https://github.com/edx/edx-platform/blob/2f12e0b4c47e99aaaf900b5288d33f902aba791d/common/lib/xmodule/xmodule/capa_module.py#L359-L407) when appropriate.

Reply:
I think then we might have a bug because of this code here: 
https://github.com/edx/edx-platform/blob/055f7323f7f0c5389aa81dd46935499310b0e258/lms/djangoapps/instructor_task/tasks_helper/grades.py#L673-L678

It's creating a dict keyed on the username, so if it gets two rows with the same username, it will only keep one. I did a quick test and this seems to be the case. 

Given that, I think the code @john2x wrote is correct, but we need to fix the report generator. This will also need some additonal thought because it means we now have multiple responses with the same block key and state and I don't know how that will interfere with existing tools. Some people were affected when the state column was removed.

Reply:
Thanks @xitij2000. You are right about the other answers overwriting the previous ones. This is actually apparent in my example output above (there should've been 6 rows). I missed that.

I can flatten the questions for the Survey report like `'Question ({})'.format(question_id)` for the meantime to fit the current expectations of the report generator. Maybe something like the following (question ids are `enjoy` and `learn`)?

```
username,...,Question enjoy,Question enjoy Answer ID,Question enjoy Answer,Question learn,Question learn Answer ID,Question learn Answer,...
john,...,Are you enjoying the course?,Y,Yes,Do you think you will learn a lot?,M,Maybe,...
```

Though for the Capa report, there doesn't seem to be question id's that can be used for the header. We could use the question label/text itself for the headers? But that might get too long.

Reply:
How much work will it be to open a fix PR for making the platform's report generator accept multiple entries? Then once we have that PR open we can ask edX + the community if it will break any tooling.

Reply:
@bradenmacdonald I don't have enough context yet to know what would be the best shape for the data, but if we want to keep the `{username1: state, username2: state}` shape from [here](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/instructor_task/tasks_helper/grades.py#L674-L678), maybe we can change it to `{username1: [state], username2: [state]}` to accommodate multiple state values? Then update the CSV generation to write one line per state. This should be a fairly quick change, but I'll try to dig a bit more to see what other parts in `edx-platform` would be affected from changing to a list of states. I can open a PR for this, but would this need a new Jira task?

Reply:
I don't think it's much work since I know exactly what to do. The only issue would be the duplicated state entries. 

One option would be to add a checkbox or similar to the UI that lets people switch between raw reports or enhanced reports. If they unselect enhanced reports (default is selected) then we only output the state in old-style reports. 

Is it worth posting this in slack? Or perhaps the forum? I'll make a PR first and link to that either way.

Reply:
@john2x Yup, that is exactly what I was going for :-) No other places would be affected other than the report, so this is a "safe" change to make.

Reply:
Thanks @xitij2000 :) So just to confirm, we'll keep the Survey report to yield multiple answers per username?

Reply:
@john2x Yeah, this will be fixed in the generate method.